### PR TITLE
Suggested ow_lander action refactoring

### PR DESCRIPTION
**Pull request by AstroStucky**
# Linked Issues:
For feedback on the approach to [OCEANWATER-904](https://babelfish.arc.nasa.gov/jira/browse/OCEANWATER-904)

# Summary of Changes
* **ow_actions** python package created to facilitate OceanWATERS lander actions. This provides a framework for 
  * Creating action for OceanWATERS
  * Easily creating actions servers or action client scripts from said actions
  * Unifying the implementation of arm actions and encapsulating arm action code behind a simple interface
* **LighSetIntensity**, **CameraCapture**, **Stow**, and **Unstow** all converted to this framework to provide examples of its usage. 
* Reduced repeated code and logic processes throughout arm action server implementation.

# Temporary Caveats
* Deactivated all other arm actions in this branch. This is only to ensure there are no collisions between the new `arm_action_server_experimental.py` (temporary name) and `arm_action_server.py` (which has been deactivated).
* The stop feature for arm actions is not yet implemented into the new framework, so `stop_action_client.py` will do nothing. The feature will be implemented in a later merge onto the OW-904 branch.

# Code Review Guide
This PR includes some code copy-pasting from `ow_lander/scripts`, much of which is only different by an indent. This section tells you what you can ignore since GitHub's diff won't catch what is the same for these files. 

## action_trajectories.py -> trajectory_planner.py 
* The class was made into a singleton called `ArmTrajectoryPlanner`, and the method `cascade_plans`,  now `_cascade_plans`, was moved into the global scope.
* An `__init__` function was added to initialize the singleton state. This made many function parameters obsolete, so wherever `move_arm`, `move_grinder`, `robot`, or `moveit_fk` were passed into a function, those parameters have been removed and their references in the method replaced with their corresponding member variables. 
* The `sever_stop` parameter was also removed, but it's functionality has not yet been replaced (see Caveats)
* `compute_forward_kinematics` method added
* `plan_arm_to_target` method added
* `check_for_stop` appeared to do nothing, so it's logic was disabled. Now always returns False. Will be removed later.

## trajectory_async_execution.py -> trajectory_executor.py
* The class was made into a singleton called `ArmTrajectoryExecutor`.
* An `__init__` function was added to initialize the singleton state. All the logic in the `connect` function was moved into here. 
* `connect` member variable removed. If a connection fails in __init__ the top level script fails with an exception. 

## utils.py -> common.py 
* Added the Singleton metaclass.

## constants.py -> constants.py (moved into ow_actions)
No changes from the original!

# Testing
## Test Stow and Unstow
1. Launch simulation into any world. `roslaunch ow atacama_y1a.launch`
2. Open 4 other terminals that all source `devel/setup.bash`:
  - **1st terminal** enter `rostopic echo /Unstow/goal`
  - **2nd terminal** enter `rostopic echo /Unstow/feedback` 
  - **3rd terminal** enter `rostopic echo /Unstow/result`
  - **4th terminal**, call the unstow action with `rosrun ow_lander unstow_client.py`
3. You should see a single goal populate the **1st terminal**, and it'll look something like 
```
header: 
  seq: 1
  stamp: 
    secs: 1916525590
    nsecs: 720000029
  frame_id: ''
goal_id: 
  stamp: 
    secs: 1916525590
    nsecs: 720000029
  id: "/unstow_client_647975_1669927495337-1-1916525590.720"
goal: 
  goal: 0.0
---
```
You should see the **2nd terminal** publish a feedback at a rapid rate (100 Hz). It'll look like
```
header: 
  seq: 4200
  stamp: 
    secs: 1916525604
    nsecs: 755000114
  frame_id: ''
status: 
  goal_id: 
    stamp: 
      secs: 1916525590
      nsecs: 720000029
    id: "/unstow_client_647975_1669927495337-1-1916525590.720"
  status: 1
  text: "This goal has been accepted by the simple action server"
feedback: 
  current: 
    x: 0.7587987870086269
    y: 0.18979331807595998
    z: 0.8811694656849518
---
```
4. The arm should move into the unstowed position, and when it arrives you should see the following goal in the **3rd terminal**
```
header: 
  seq: 4
  stamp: 
    secs: 1916525605
    nsecs: 444999933
  frame_id: ''
status: 
  goal_id: 
    stamp: 
      secs: 1916525590
      nsecs: 720000029
    id: "/unstow_client_647975_1669927495337-1-1916525590.720"
  status: 3
  text: "Unstow arm trajectory succeeded"
result: 
  final: 
    x: 0.7355859052905875
    y: 0.2552613456452144
    z: 0.9050746780690109
---
```
5. In any terminal, call stow `rosrun ow_lander stow_client.py` and verify the arm returns to the stowed position. No need to check goal/feedback/result messages here. 
6. Repeat stows and unstows as many times as you'd like. 

## Test Camera Capture
The following can be done in the same simulation world. No need to check action goal/feedback/result topics. 
1. Move mast to look at ground `rosrun ow_lander antenna_pan_tilt_action_client.py 1.0 0.5`
2. Bring up the Image View container in RQT.
3. Capture a picture using whatever exposure you'd like `rosrun ow_lander camera_capture_client.py 0.1`
4. Verify an image appeared in RQT.

## Test Light Set Intensity
1. Move mast to project its lights on the lander body `rosrun ow_lander antenna_pan_tilt_action_client.py 0.0 1.0`
2. Change light intensity as you like `rosrun ow_lander light_set_intensity_client.py`
3. Verify the corresponding mast light change intensity levels in the Gazebo sim. 

## Test Client Help messages
Try the following commands and verify the resulting help messages makes sense. 
```rosrun ow_lander unstow_client.py --help```
```rosrun ow_lander stow_client.py --help```
```rosrun ow_lander camera_capture_client.py --help```
```rosrun ow_lander light_set_intensity_client.py --help```

### CODE REVIEW

**Comment by AstroStucky** [comment]
WOW. Sorry for the lengthy PR. The Testing section looks long, but I promise it shouldn't take very long. Everything is there to help you out as much as possible because this will be a slightly tricky PR to review.

**Comment by kmdalal** [comment]
The tests all passed for me.

My first comment on the code is that I find the new directory structure a bit complicated and confusing, i.e. `/ow_lander/src/ow_actions/actions`:
 - We've used `ow_` to prefix only repositories and packages, so a third use is confusing.
 - Having 'actions' under 'ow_actions' is a bit confusing.
 - I do like the separation of 'scripts' and 'src', but what I'll suggest from what I see so far is either putting all Python files flat under 'src', or having something like `src/actions' and 'src/common'.
 - I don't understand the empty __init__.py.

**Comment by AstroStucky** [comment]
> * Python is now compiled (great) but the instructions don't mention this.  Now, a build is actually useful after updating Python scripts.

They are? I am honestly not sure how to tell. 😅 Perhaps this is a result of using a setup.py script?

> * The following command from the instructions fails for reasons I don't understand.  The same script invocation works fine when done without 'rosrun'.
This is my mistake. All script invocations in my instructions should be done with a `rosrun`. It has been fixed in the testing procedure. 

As I understand, rosrun simply executes the name of an executable script provided by the second argument so long as it is found in the package provided by the first argument. We have actually always had this facility available to us, and I find it much nicer than providing a full script path or changing the directory.

> * The two antenna positions specified are the same: `rosrun ow_lander antenna_pan_tilt_action_client.py 1.0 0.5`

That is another mistake in the procedure. And I have fixed it.

**Comment by AstroStucky** [comment]
> My first comment on the code is that I find the new directory structure a bit complicated and confusing, i.e. `/ow_lander/src/ow_actions/actions`:
> 
> * We've used `ow_` to prefix only repositories and packages, so a third use is confusing.
> * Having 'actions' under 'ow_actions' is a bit confusing.

This largely comes from the ROS [PyStyleGuide](http://wiki.ros.org/PyStyleGuide#Package.2FModule_Names_.28__init__.py_files.29). I took some liberty in naming of the python package (**ow_actions**), only because I thought it was appropriate seeing as it only handles lander actions. The PyStyleGuide actually encourages using the same name as the overhead ROS package, such that **ow_actions** would become **ow_lander**. I don't imagine that would address your concerns, but it would actually be more ROS-like. 
**Source** http://wiki.ros.org/PyStyleGuide#Package.2FModule_Names_.28__init__.py_files.29

> * I do like the separation of 'scripts' and 'src', but what I'll suggest from what I see so far is either putting all action server implementation flat under 'src', ...

That is what I intend to do, but it is hard to see in this branch so far because there are still many large python scripts leftover in `scripts/` directory that will either be deleted (i.e. the ROS service implementation) or still need to be converted to this new format.

> ... or having something like `src/actions' (or src/servers) and 'src/common' (or src/action_support).  Other good names are possible.

I am open to a new file hierarchy, but I would like to suggest something a little different that follows the PyStyleGuide. See a follow-up comment for that, since this one is getting a little too verbose. 

> * I don't understand the empty underscore-init-underscore.py.

This is actually just how one must define a Python package. Every Python package must have an `__init__.py` file in its directory to be imported into a script or module, and they are most typically empty unless you need to define an API  or do something specialized. This rule also applies to sub-directories of the package, which are treated like sub-packages of their own. I admit that I don't have a solid grasp of best practices around `__init__.py` files, which is why I tend to default to leaving them empty. 
**Source** https://docs.python.org/3/tutorial/modules.html#packages

**Comment by AstroStucky** [comment]
Here's an alternative file hierarchy I want to suggest in lieu of Mike's comment. 
```
`-- ow_lander
    `-- src
        `-- ow_lander
            |-- action
            |   |-- servers
            |   |   |-- __init__.py
            |   |   |-- camera_capture.py
            |   |   |-- light_set_intensity.py
            |   |   |-- stow.py
            |   |   |-- unstow.py
            |   |-- __init__.py
            |   |-- arm.py
            |   |-- node_helper.py
            |   |-- server.py
            |   |-- trajectory_executor.py
            |   `-- trajectory_planner.py
            |-- __init__.py
            |-- common.py
            |-- constants.py
            |-- subscribers.py
```

**Comment by kmdalal** [comment]
Taking a step back, the directory reorganization came as a surprise.  It's not mentioned in the Jira, and I thought the goal of this effort was purely a code refactoring, mainly to eliminate a lot of repetition.  The immediate practical motivation is implementing a number of new actions needed soon for the OWLAT unification.

While the latest proposed directory structure follows the ROS style guide, does it have any practical benefit?  Does changing the directory structure at all (e.g. leaving all Python code in `scripts`) have any practical benefit? 

As for the style recommended in the ROS wiki, it does not look appealing to me -- it looks verbose and confusing.   We don't need to follow that wiki, especially for a system as relatively small and simple as ours.

I would like to see directory reorg (if desired) either done as a separate effort, or have just a small directory change included with this one, i.e. splitting 'scripts' and adding 'src'.  But it looks like having 'src' may imply a bunch of other conventions so I'm skeptical of that.

**Comment by AstroStucky** [comment]
> It's not mentioned in the Jira, and I thought the goal of this effort was purely a code refactoring, mainly to eliminate a lot of repetition. 

I did not think it needs to. The purpose was to improve the structure of the code without impacting the outward functionality. Aside from a few improvements to logging, that is what this is. Having everything in a scripts directory, including things that are not scripts, is a poor structure.

> While the latest proposed directory structure follows the ROS style guide, does it have any practical benefit? Does changing the directory structure at all (e.g. leaving all Python code in scripts) have any practical benefit?

We conform to the ROS python style guide. That in and of itself is a practical benefit, particularly for an open source project. An additional benefit is that I can glance at the directory tree and know what and where functionality for lander operations is implemented. I could not do that with the scripts directory.

**Comment by kmdalal** [comment]
Okay, since I'm neither a strong Python developer nor a ROS expert, and since this area of the system will be owned by you for the near future, feel free to lay it out in a way that reasonably follows ROS conventions, and (more importantly in my opinion) looks good and makes sense to _you_.  I'll defer further comments to @anthonykoutroulis .

File: ow_lander/src/ow_actions/arm.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/arm.py:4]: A short explanation of what this file is for would be helpful, especially since the filename is so simple.  On that note, should this file be named ArmActionMixin.py?  (I don't know Python naming conventions).

File: ow_lander/src/ow_actions/common.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/common.py:17]: This function seems rather specific, compared to the rest of the file.   Why not make it applicable to any joint?

File: ow_lander/src/ow_actions/trajectory_executor.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/trajectory_executor.py:66]: typo

File: ow_lander/src/ow_actions/arm.py
Comment (AstroStucky) [reply ow_lander/src/ow_actions/arm.py:4]: I could stand to add more file descriptions.

Aside from requiring snake_case and brevity where possible there aren't much in the way of module naming conventions. `arm_action_mixin.py` works fine for me. 

Alternatively, I could put all of the arm interface classes (ArmTrajectoryPlanner, ArmTrajectoryExecutor, and ArmActionMixin) under a single module named `arm.py`. Thoughts?

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/trajectory_planner.py:53]: Anything not used seems better removed.

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/trajectory_planner.py:56]: In the interest of completing this refactoring phase, suggest dealing with this later.  Arm stopping seems a bit complicated and worthy of separate attention.

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/trajectory_planner.py:37]: Remove?

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/trajectory_planner.py:571]: Could be reformatted for easier reading.

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (kmdalal) [inline ow_lander/src/ow_actions/trajectory_planner.py:877]: Minor: shorten line

**Comment by AstroStucky** [comment]
To make rename/modify detection better, some files from `scripts` that were copied to `ow_actions` were reverted back to their original indentation size (4 space), and their originals in `scripts` removed.  For `common.py` and `constants.py` this worked great; however, for `trajectory_executor.py` and `trajectory_planner.py` the similarity indices calculated by git are below the treshold (50%) to be considered rename/mods. 

There is no way to alter the similarity threshold in the GitHub PR interface, but if I reviewer is so inclined they can do so on their own machine. While on the `suggested-action_refactor` branch type the following to generate a diff that considers both `trajectory_executor.py` and `trajectory_planner.py` rename and modifications of `trajectory_async_execution.py` and `action_trajectories.py` respecitvely. 
```
git diff OCEANWATER-904_action_server_refactor --stat -M30%
```

File: ow_lander/src/ow_actions/common.py
Comment (AstroStucky) [reply ow_lander/src/ow_actions/common.py:17]: This is left over from a previous implementation (`utils.py`). It may be overly specific, but I would prefer not change more than I have to for this refactor.

**Comment by AstroStucky** [comment]
> ```
> git diff OCEANWATER-904_action_server_refactor -M28%
> ```
I just wanted to share a more straight forward way to do this for those files individually.
```
git diff OCEANWATER-904_action_server_refactor:./scripts/action_trajectories.py src/ow_lander/trajectory_planner.py
```
and
```
git diff OCEANWATER-904_action_server_refactor:./scripts/action_trajectories.py src/ow_lander/trajectory_planner.py
```

**Comment by AstroStucky** [comment]
Suggested changes have been incorporated.

File: ow_lander/src/ow_lander/trajectory_executor.py
Comment (kmdalal) [inline ow_lander/src/ow_lander/trajectory_executor.py:106]: typo

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (kmdalal) [reply ow_lander/src/ow_actions/trajectory_planner.py]: I've tested Stow and Unstow and find that:
 - The Stop action no longer works
 - Arm faults behave as though "arm_continues_in_fault" is selected.  I.e. the unfaulted joints keep moving.

Again I'd suggest a rework of this functionality be separate from refactoring, and even pushed into Release 12 unless something is broken.

File: ow_lander/src/ow_lander/arm.py
Comment (kmdalal) [inline ow_lander/src/ow_lander/arm.py:1]: I'll suggest arm_action_mixin.py as the name of this file.

File: ow_lander/src/ow_lander/trajectory_planner.py
Comment (kmdalal) [inline ow_lander/src/ow_lander/trajectory_planner.py:1]: A lot of refactoring seems possible within this file, but I imagine some of it must wait until more arm actions are brought into the new package structure.

File: ow_lander/src/ow_lander/actions/stow.py
Comment (AstroStucky) [reply ow_lander/src/ow_lander/actions/stow.py:12]: You're not wrong, they are very alike. The definition of execute_action has already changed a little in the stop action branch, whic I am wrapping up now. 

I could try and experiment with doing something like this in that branch, but I am hesitant to hide away too much action implementation behind multiple levels of parent classes because at a certain point it makes seeing what the action does at a glance harder.

File: ow_lander/scripts/unstow_client.py
Comment (mogumbo) [inline ow_lander/scripts/unstow_client.py:13]: There's new code like this with 4-space indentation. It would be good to make it 2-space to be consistent with the bulk of the code in this PR.

**Review by mogumbo — CHANGES_REQUESTED** [review]

File: ow_lander/src/ow_lander/trajectory_planner.py
Comment (AstroStucky) [reply ow_lander/src/ow_lander/trajectory_planner.py:1]: It is partly that, but mostly that there is not enough time to include a refactor of this at present and also make our deadline for command unification.

File: ow_lander/launch/spawn.launch
Comment (AstroStucky) [reply ow_lander/launch/spawn.launch:103]: The main reason I did it this way is so I can still review the old implementation that is in `arm_action_servers.py` as I go. How does it sound to you if I rename this to `arm_action_servers.py` and rename the old implementation to `arm_action_servers_old.py`, which will then be removed entirely in a later PR, but before being merged into noetic-devel?

File: ow_lander/launch/spawn.launch
Comment (AstroStucky) [reply ow_lander/launch/spawn.launch:106]: I cannot remember why I even did this since pan/tilt and dock ingest sample are not broken in this branch. They have been uncommented.

File: ow_lander/scripts/unstow_client.py
Comment (AstroStucky) [reply ow_lander/scripts/unstow_client.py:13]: Done.

File: ow_lander/launch/spawn.launch
Comment (AstroStucky) [reply ow_lander/launch/spawn.launch:103]: On second thought, I'll just remove the old `arm_action_servers.py` entirely from this branch. I'll keep it's old implementation on file locally for my own reference, but there's no need to keep it in this branch.

Comment (AstroStucky) [reply ow_lander/launch/spawn.launch:103]: `arm_action_servers.py` is now the one and only implementation of this file, and it is the new experimental version.

File: ow_lander/src/ow_actions/trajectory_planner.py
Comment (AstroStucky) [reply ow_lander/src/ow_actions/trajectory_planner.py]: Just reiterating our discussion from this morning: Yes, things are broken in this branch, namely: stop arm action and stop arm actions on fault. I have an implementation ready for review once this review is approved for a merge. Followed by that will be an implementation of all of our ow_lander actions into this new framework.

File: ow_lander/src/ow_lander/actions/stow.py
Comment (kmdalal) [reply ow_lander/src/ow_lander/actions/stow.py:12]: In my opinion it would be more helpful to see at a glance that stow and unstow are almost identical!   Refactoring by nature often adds abstraction.

**Comment by AstroStucky** [comment]
```
$ git diff noetic-devel --shortstat -M28%
 29 files changed, 1117 insertions(+), 1844 deletions(-)
```
Now this is looking like a real code reduction!

**Comment by kmdalal** [comment]
The latest version of this branch has processes crashing on simulation startup for me, even after a reboot, clean build, resourcing environment, and several tries.    Here are the two crashes I'm getting.
```
[ArmActionServers-18] process has died [pid 3130, exit code 1, cmd bash -c sleep 10.0; $0 $@ /usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/arm_action_servers.py __name:=ArmActionServers __log:=/home/kdalal/.ros/log/ca2799ae-7755-11ed-a756-9ffc1df42e09/ArmActionServers-18.log].
log file: /home/kdalal/.ros/log/ca2799ae-7755-11ed-a756-9ffc1df42e09/ArmActionServers-18*.log
[AntennaPanTiltAction-19] process has died [pid 3133, exit code 1, cmd bash -c sleep 10.0; $0 $@ /usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/antenna_pan_tilt_action_server.py __name:=AntennaPanTiltAction __log:=/home/kdalal/.ros/log/ca2799ae-7755-11ed-a756-9ffc1df42e09/AntennaPanTiltAction-19.log].
log file: /home/kdalal/.ros/log/ca2799ae-7755-11ed-a756-9ffc1df42e09/AntennaPanTiltAction-19*.log


Traceback (most recent call last):
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/path_planning_commander.py", line 17, in <module>
    import activity_full_digging_traj
  File "/usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/activity_full_digging_traj.py", line 8, in <module>
    import constants
ModuleNotFoundError: No module named 'constants'
[path_planning_commander-16] process has died [pid 8601, exit code 1, cmd /usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/path_planning_commander.py __name:=path_planning_commander __log:=/home/kdalal/.ros/log/f65620b2-7756-11ed-a756-9ffc1df42e09/path_planning_commander-16.log].
log file: /home/kdalal/.ros/log/f65620b2-7756-11ed-a756-9ffc1df42e09/path_planning_commander-16*.log
[Msg] Loading world file [/usr/local/home/kdalal/ow/src/ow_europa/worlds/atacama_y1a.world]
ERROR 1: PROJ: proj_create_operations: Source and target ellipsoid do not belong to the same celestial body
ERROR 6: Cannot find coordinate operations from `PROJCRS["ORTHOGRAPHIC EUROPA",BASEGEOGCRS["GCS_EUROPA",DATUM["D_EUROPA",ELLIPSOID["EUROPA",1561500,0,LENGTHUNIT["metre",1,ID["EPSG",9001]]]],PRIMEM["Reference_Meridian",0,ANGLEUNIT["degree",0.0174532925199433,ID["EPSG",9122]]]],CONVERSION["unnamed",METHOD["Orthographic",ID["EPSG",9840]],PARAMETER["Latitude of natural origin",-32.6500015,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8801]],PARAMETER["Longitude of natural origin",332.5400085,ANGLEUNIT["degree",0.0174532925199433],ID["EPSG",8802]],PARAMETER["False easting",0,LENGTHUNIT["metre",1],ID["EPSG",8806]],PARAMETER["False northing",0,LENGTHUNIT["metre",1],ID["EPSG",8807]]],CS[Cartesian,2],AXIS["easting",east,ORDER[1],LENGTHUNIT["metre",1,ID["EPSG",9001]]],AXIS["northing",north,ORDER[2],LENGTHUNIT["metre",1,ID["EPSG",9001]]]]' to `EPSG:4326'
[Err] [Dem.cc:236] EXCEPTION: Unable to transform terrain coordinate system to WGS84 for coordinates (0,0)
```

File: ow_lander/src/ow_lander/actions/stow.py
Comment (AstroStucky) [reply ow_lander/src/ow_lander/actions/stow.py:12]: Yea I see your point. At the risk of getting too in the weeds on this thread, all of the arm actions can be categorized into 4 general types that could all simply be different specializations of an ArmActionMixin.
1. **Moves arm direct to target joint values** - Stow and Unstow
2. **Retrieve a complex trajectory directly from trajectory_planner.py** - Grind, Dig Linear, Dig Circular, Deliver, and Discard
3. **Modifies existing joint values** - Arm Move Joint and Arm Move Joints
4. **Specialized behavior (may simply inherit ArmActionMixin directly)** - Stop and Guarded Move (which _might_ work as a **type 2** with extra bits added)

You are suggesting I make a shared base class for **type 1**. That sounds fine to me. What are your thoughts on doing this for the other types that I have listed? Keep in mind that command unification may break some of these types, and add brand new ones.

**Comment by AstroStucky** [comment]
The crash of `arm_action_servers.py` has been fixed. I just forgot to change import statements.

`path_planning_commander.py` is actually an implementation specific to Arm ROS Services. It looks like it relies on the now significantly modified `trajectory_async_exectuion.py`, and I am wondering if it is worth bothering to fix?

File: ow_lander/src/ow_lander/actions/stow.py
Comment (kmdalal) [reply ow_lander/src/ow_lander/actions/stow.py:12]: This sounds like a good analysis and candidate for refactoring, with (1) being the easiest.  In the interest of a timely release, I'll suggest postponing this level of refactoring to after the unified commands are all implemented.

**Comment by kmdalal** [comment]
> The crash of `arm_action_servers.py` has been fixed. I just forgot to change import statements.

Good, I'll pull and retry.

> `path_planning_commander.py` is actually an implementation specific to Arm ROS Services. It looks like it relies on the now significantly modified `trajectory_async_exectuion.py`, and I am wondering if it is worth bothering to fix?

I think it should either be fixed, or better removed from the launch if it's not (yet) needed.

**Comment by kmdalal** [comment]
I'm still getting the following crash.  After that's resolved I'm ready to approve this unless there are more changes coming...
```
[path_planning_commander-16] process has died [pid 15680, exit code 1, cmd /usr/local/home/kdalal/ow/src/ow_simulator/ow_lander/scripts/path_planning_commander.py __name:=path_planning_commander __log:=/home/kdalal/.ros/log/c9eb81a4-77f9-11ed-817c-99d9f5ad472f/path_planning_commander-16.log].
log file: /home/kdalal/.ros/log/c9eb81a4-77f9-11ed-817c-99d9f5ad472f/path_planning_commander-16*.log
```

**Comment by AstroStucky** [comment]
path_planning_commander has been disabled in `spawn.launch`. It could be fixed instead, but I think we should discuss what the ultimate fate of our lander ROS services will be first.

**Comment by kmdalal** [comment]
> path_planning_commander has been disabled in `spawn.launch`. It could be fixed instead, but I think we should discuss what the ultimate fate of our lander ROS services will be first.

I didn't realize this was the Services server.  We decided a few months ago to eliminate services altogether and remove them from our code, whenever convenient.  I would not include that in this refactor though.

**Review by kmdalal — APPROVED** [review]
Approving this for now, but re-request if significant changes are made.  Good work!

**Comment by anthonykoutroulis** [comment]
In the testing instructions, terminals 1 and 3 are instructed to echo the same topic. What topic should terminal 3 be observing?

**Comment by anthonykoutroulis** [comment]
The light intensity action client script's help suggests that `{left, right}` arguments are optional and can be omitted, but running without a positional argument (ie: `rosrun ow_lander light_set_intensity_action_client.py 0.0`) throws an error:
```
usage: light_set_intensity_action_client.py [-h] [{left,right}] [intensity]
light_set_intensity_action_client.py: error: argument name: invalid choice: '0.0' (choose from 'left', 'right')
```

**Comment by AstroStucky** [comment]
> The light intensity action client script's help suggests that `{left, right}` arguments are optional and can be omitted, but running without a positional argument (ie: `rosrun ow_lander light_set_intensity_action_client.py 0.0`) throws an error:
> 
> ```
> usage: light_set_intensity_action_client.py [-h] [{left,right}] [intensity]
> light_set_intensity_action_client.py: error: argument name: invalid choice: '0.0' (choose from 'left', 'right')
> ```

This is less a bug, and more a misleading help message. By using `nargs='?'` for all of our action client arguments we're essentially telling argparse to consider everything after the final provided argument to be optional. Or to put it another way, yes everything is optional, but you also cannot skip any arguments in the sequence. Your command throws an error because the first argument in the sequence "0.0" is interpreted by argparse as the 'name' argument rather than the 'intensity' argument.

I'll use dig linear for a more illustrative example of the issue. If I wanted to dig a trench using the default parameters, I would call it without any arguments.
```
rosrun ow_lander dig_linear_action_client.py
```
If I then want to dig a couple centimeters deeper, but otherwise keep the other parameters the same, I would actually have to manually type in all of the defaults up to the parameter I actually want to change. 
```
rosrunw ow_lander dig_linear_action_client.py 1.46 0.0 0.03
```

I don't like having to do it this way. I think the scripts could be improved, but I did not want to do that improvement with this refactor work, and a solution may call for some team discussion.

**Comment by anthonykoutroulis** [comment]
> This is less a bug, and more a misleading help message...
Makes sense to me, it was a pretty minor comment. Definitely something we can address later if needed.

Okay, all looks good to merge to me.

**Review by anthonykoutroulis — APPROVED** [review]

**Pull request merged**
